compiler/rustc_target/src/target_features.rs

```diff
@@ -904,27 +718,108 @@ impl Target {
+        // defined by target features. When that is the case, those target features must be
+        // "forbidden" in the list above to ensure that there is a consistent answer to the
+        // questions "which ABI is used".
+        match &*self.arch {
+            "x86" => {
+                // We support 2 ABIs, hardfloat (default) and softfloat.
+                if self.has_feature("soft-float") {
+                    NOTHING
+                } else {
+                    // Hardfloat ABI. x87 must be enabled.
+                    (&["x87"], &[])
+                }
+            }
+            "x86_64" => {
+                // We support 2 ABIs, hardfloat (default) and softfloat.
+                if self.has_feature("soft-float") {
+                    NOTHING
+                } else {
+                    // Hardfloat ABI. x87 and SSE2 must be enabled.
+                    (&["x87", "sse2"], &[])
+                }
+            }
+            "arm" => {
+                // We support 2 ABIs, hardfloat (default) and softfloat.
+                if self.has_feature("soft-float") {
+                    NOTHING
+                } else {
+                    // Hardfloat ABI. x87 must be enabled.
+                    (&["fpregs"], &[])
+                }
+            }
+            "aarch64" | "arm64ec" => {
+                match &*self.abi {
+                    "softfloat" => {
+                        // This is not fully correct, LLVM actually doesn't let us enforce the softfloat
+                        // ABI properly... see <https://github.com/rust-lang/rust/issues/134375>.
+                        // FIXME: should be forbid "neon" here? But that would be a breaking change.
+                        NOTHING
+                    }
+                    "uwp" | "llvm" | "macabi" | "sim" | "ilp32" | "" => {
+                        // Hardfloat ABI. neon and fp-armv8 must be enabled.
+                        (&["neon", "fp-armv8"], &[])
+                    }
+                    _ => unreachable!(),
+                }
+            }
+            "riscv32" | "riscv64" => {
+                match &*self.llvm_abiname {
+                    "ilp32d" | "lp64d" => {
+                        // Requires f and d, incompatible with e.
+                        (&["f", "d"], &["e"])
+                    }
+                    "ilp32f" | "lp64f" => {
+                        // Requires f, incompatible with e.
+                        (&["f"], &["e"])
+                    }
+                    "ilp32e" => {
+                        // ilp32e is incompatible with features that need aligned load/stores > 32 bits,
+                        // like `d`. And of course `e` is required.
+                        (&["e"], &["d"])
+                    }
+                    "lp64e" => {
+                        // Embedded ABI, requires e.
+                        (&["e"], &[])
+                    }
+                    "ilp32" | "lp64" => NOTHING,
```

Comment:
```suggestion
                    "ilp32" | "lp64" => {
                        // Incompatible with e.
                        (&[], &["e"])
                    }
```

Reply:
Ah right the old logic treated all the non-e ABIs the same, we should keep that.

The old logic also allowed disabling e when using an e ABI, so we should probably keep that, too.